codegen/src/PulumiPackageInfo.scala

```diff
@@ -1,20 +1,255 @@
 package besom.codegen
 
-import besom.codegen.metaschema.ConstValue
+import besom.codegen.Utils.{FunctionName, PulumiPackageOps, isMethod}
+import besom.codegen.metaschema.{FunctionDefinition, *}
 import besom.codegen.{PackageVersion, SchemaName}
 
+import scala.collection.mutable.ListBuffer
+import scala.util.matching.Regex
+
 type EnumInstanceName = String
 
 case class PulumiPackageInfo(
   name: SchemaName,
   version: PackageVersion,
-  enumTypeTokens: Set[String],
-  objectTypeTokens: Set[String],
   providerTypeToken: String,
-  resourceTypeTokens: Set[String],
   moduleToPackageParts: String => Seq[String],
   providerToPackageParts: String => Seq[String],
-  enumValueToInstances: Map[PulumiToken, Map[ConstValue, EnumInstanceName]]
+  enumTypeTokens: Set[String],
+  objectTypeTokens: Set[String],
+  resourceTypeTokens: Set[String],
+  enumValueToInstances: Map[PulumiToken, Map[ConstValue, EnumInstanceName]],
+  parsedTypes: Map[PulumiDefinitionCoordinates, (TypeDefinition, Boolean)],
+  parsedResources: Map[PulumiDefinitionCoordinates, (ResourceDefinition, Boolean)],
+  parsedFunctions: Map[PulumiDefinitionCoordinates, (FunctionDefinition, Boolean)]
+)(
+  private[codegen] val pulumiPackage: PulumiPackage,
+  private[codegen] val providerConfig: Config.Provider
 ) {
+  import PulumiPackageInfo.*
+
   def asPackageMetadata: PackageMetadata = PackageMetadata(name, Some(version))
+
+  def parseMethods(
```

Comment:
the parsing logic moved from other places (mostly SchemaProvider and PulumiPackageOps)